File: workflow/scripts/02_barcode_qc.py
import matplotlib.pyplot as plt
import HTSeq
import itertools
from collections import Counter
import pandas as pd
import sys
import json

#NOTE
#Current script is set up for two barcoding strategies - if additional are required they 
# must be added below. Current accepted formats include:
#NNNNNCACTGTNNNN = total len 15, seq len 6, position [5:11]
#NNNTGGCNN = total len 9, seq len 4, position [3:7]

#args
sample_manifest = sys.argv[1] #"samples_corrected.tsv"
multiplex_manifest = sys.argv[2] #"multiplex_corrected.tsv"
barcode_input = sys.argv[3]
output_dir = sys.argv[4] #location input barcode counts, output barcode plots and summary
mismatch = sys.argv[5] #2

def CreateSampleDicts(df_m,df_s):
    s_dict={}
    m_dict={}

    #create dict that maps multiplex: sampleid : barcode
    for multi_name in df_s['multiplex'].unique():
        sub_df = df_s[df_s['multiplex']==multi_name]
        s_dict[multi_name]={}

        #each row maps to a sample:barcode and sample:group
        for index,row in sub_df.iterrows():
            sample_name = row['sample']
            s_dict[multi_name][sample_name]=row['barcode']

    #creat dict that maps mulitplex: filename
    #remove .fastq.gz from filename
    m_dict = dict(zip(df_m.multiplex, df_m.file_name.str.replace('.fastq.gz','')))

    return(m_dict,s_dict)

#Read sample/multiplex manifests, create dicts
df_multiplex = pd.read_csv(multiplex_manifest,sep="\t")
df_samples = pd.read_csv(sample_manifest,sep="\t")
(multiplex_dict,samp_dict) = CreateSampleDicts(df_multiplex,df_samples)

#for each multiplexed sample: fastq file
for k,v in multiplex_dict.items():
    #create expected bc list
    bc_exp=[]
    for k2,v2 in samp_dict[k].items():
        bc_exp.append(v2.replace('N',''))

    #check the number of mismatches requested is possible given the expected barcodes
    for a, b in itertools.combinations(bc_exp, 2):
      
        #zip two barcodes and compare each letter
        compare=zip(a,b)
        diff_list=[]
        for i,j in compare:
            #if the letters are a mismatch, add to list
            if i!=j:
                diff_list.append(j)
            #if the mismatch list is not > than the mismatch value requested, print error and fail
        if(len(diff_list)<1 + int(mismatch)):
            print('The number of differences ({}) between barcodes {} and {} is less than or equal to the number of mismatches requested ({})'. format(len(diff_list),a,b,mismatch))
            
            sys.exit('Barcode strategy requires differences between barcodes is greater than mismatch allowance')

    #generate list of possible barcodes variations
    bc_mutants={}
    nuc_list = ['A','T','C','G','N']

    # for each expected barcode
    for bc in bc_exp:
        #vary the string by 1bp, 
        #push str to dict with expected value as the dict key
        for i in range(0,len(list(bc))):
            for nuc in nuc_list:
                tmp = list(bc)
                tmp[i]=nuc
                bc_mutants["".join(tmp)]=bc

                #if more than 1 mismatch is allowed, repeat variation
                if(int(mismatch)==2):
                    for j in range(0,len(list(bc))):
                        for nuc in nuc_list:
                            tmp2 = list("".join(tmp))
                            tmp2[j]=nuc
                            bc_mutants["".join(tmp2)]=bc
    
    # read list of counts:barcodes
    fastq_df = pd.read_csv(barcode_input,header=None)
    fastq_df[0] = fastq_df[0].str.lstrip() #remove leading spaces
    fastq_df[['counts','bc']] = fastq_df[0].str.split(expand=True)#split cols
    fastq_df["counts"] = pd.to_numeric(fastq_df["counts"])
    fastq_df=fastq_df.drop([0], axis=1) #remove merged col

    #for each barcode counted
    bc_dict = {}
    for index, row in fastq_df.iterrows():
        barcode=row['bc']

        #if barcode matches mutant, change barcode to expected value
        if barcode in bc_mutants:
          barcode_update = barcode
          barcode = bc_mutants[barcode_update]

        #if barcode isn't in dict, add it
        if k not in bc_dict: 
            bc_dict[k]={} 

        #create barcode:sample:count 
        #add barcode in dict count + row count from text file
        bc_dict[k][barcode] = bc_dict[k].get(barcode, 0) + row['counts']

    #select top 10 barcodes for each sample
    top_dict = dict(Counter(bc_dict[k]).most_common(10))
    
    #compare observed bc list with expected bc list, write output to text
    bc_obs=[]
    for k2,v2 in top_dict.items():
        bc_obs.append(k2)

    # if the expected list matches with top 10, print data
    # otherwise print error messages    
    check =  all(item in bc_obs for item in bc_exp)
    if check is True:
        file_save = output_dir + '/' + k + '_barcode.txt'
        f = open(file_save,"w+")
        f.write("\n* SampleID {}\n".format(k))
        f.write("\t + Number of mismatches allowed {}\n".format(mismatch))
        f.write("\t + The top barcodes identified {} include the expected barcodes {}\n\n".format(bc_obs, bc_exp))
        f.write("\t + List of top barcodes:counts \n")
        f.write("\t\t + " + json.dumps(top_dict))
        f.close()    
    else :
        file_save = output_dir + k + '_barcode_errors.txt'
        f = open(file_save,"w+")
        f.write("The top barcodes identified {} were not congruent with expected barcode list {}. Review associated img for more information.".format(bc_obs, bc_exp)) 
        f.close()

    # print barplot for top barcodes
    file_save = output_dir + '/' + k + '_barcode.png'
    plt.bar(*zip(*top_dict.items()))
    plt.suptitle('Top Barcodes: ' + k + '\n Number of mismatches allowed: ' + str(mismatch))
    plt.xticks(rotation='45', fontsize=6)
    for k,v in top_dict.items():
        plt.text(x=k , y =v+1 , s=str(v), color = 'black', fontweight='bold')
    plt.tight_layout()
    plt.savefig(file_save)

File: workflow/scripts/02_barcode_qc.R
# Purpose:
  # Current script is set up for two barcoding strategies; if additional are required they must be added below. 
  # Current accepted formats include:
  #  - NNNNNCACTGTNNNN = total len 15, seq len 6, position [6:11]
  #  - NNNTGGCNN = total len 9, seq len 4, position [4:7]
# Input
  # to create input, run counts
  # gunzip -c  /data/CCBR_Pipeliner/iCLIP/test/fastq/test_1.fastq.gz | 
  # sed -n '/@/{n;p}' | awk '{print substr($0, 4, 4);}' | sort -n | uniq -c > test.barcode_counts.txt

library(ggplot2)
library(argparse)
library(DescTools)

#set args
parser <- ArgumentParser()
parser$add_argument("-s","--sample_manifest", dest="sample_manifest", required=TRUE, help="/path/to/samples.tsv")
parser$add_argument("-ba","--multiplex_manifest", dest="multiplex_manifest", required=TRUE, help="/path/to/multiplex.tsv")
parser$add_argument("-bu","--barcode_input", dest="barcode_input", required=TRUE, help="/path/to/_barcode_counts.txt")
parser$add_argument("-o","--output_dir", dest="output_dir", required=TRUE, help="dir for barcode plots and summary")
parser$add_argument("-m","--mismatch", dest="mismatch", required=TRUE, help="number of mismatches allowed")
parser$add_argument("-i","--mpid", dest="mpid", required=TRUE, help="sampleid")
parser$add_argument("-q","--qc_dir", dest="qc_dir", required=TRUE, help="dir for failed QC manifests")

args <- parser$parse_args()
sample_manifest = args$sample_manifest
multiplex_manifest = args$multiplex_manifest
barcode_input = args$barcode_input
output_dir = args$output_dir
mismatch = as.integer(args$mismatch)
mpid = args$mpid
qc_dir = args$qc_dir

#test input
testing="N"
if(testing=="Y"){
  sample_manifest  = "~/../../Volumes/RBL_NCI/Wolin/mov10_par_Y_r2_01062023/manifests/samples.tsv"
  multiplex_manifest = "~/../../Volumes/RBL_NCI/Wolin/mov10_par_Y_r2_01062023/manifests/multiplex.tsv"
  barcode_input = "~/../../Volumes/RBL_NCI/Wolin/mov10_par_Y_r2_01062023/sam_testing/barcode_counts.txt"
  output_dir = "~/../../Volumes/RBL_NCI/Wolin/mov10_par_Y_r2_01062023/sam_testing"
  mismatch = 1
  mpid="CLIP"
}

#Read sample/multiplex manifests
df_multiplex = read.table(multiplex_manifest,header=TRUE,sep=",")
df_samples = read.table(sample_manifest,header=TRUE,sep=",")
df_counts = read.table(barcode_input,header=FALSE)
colnames(df_counts)=c("counts","barcode")

#create list of expected barcodes
bc_exp=gsub("N","",subset(df_samples,multiplex==mpid)$barcode)

# determine the length of the barcodes expected and set number of barcodes observed
if (length(bc_exp)<10){
  number_of_barcodes_observed=10
} else if (length(bc_exp)<20) {
  number_of_barcodes_observed=20
} else{
  number_of_barcodes_observed=as.numeric(length(bc_exp))+10
}

#Determine if the variance between any two barcodes are above the threshold of mismatches
string.diff.ex<-function(a="ATTCGAN",b="attTGTT",exclude=c("n","N","?"),ignore.case=TRUE){
  if(nchar(a)!=nchar(b)) stop("Lengths of input strings differ. Please check your input.")
  if(ignore.case==TRUE){
    a<-toupper(a)
    b<-toupper(b)
  }
  diff.a<-unlist(strsplit(a,split=""))
  diff.b<-unlist(strsplit(b,split=""))
  diff.d<-rbind(diff.a,diff.b)
  for(ex.loop in 1:length(exclude)){
    diff.d<-diff.d[,!(diff.d[1,]==exclude[ex.loop]|diff.d[2,]==exclude[ex.loop])]
  }
  differences<-sum(diff.d[1,]!=diff.d[2,])
  return(differences)
}

#determine all expected barcode comparisons
barcode_combo = as.data.frame(CombSet(bc_exp,2,repl=FALSE))
error_message = ""

#for each comparison determine the number of strings that are diff - 
#if less than the mismatch threshold then document error
for (rowid in rownames(barcode_combo)){
  val = string.diff.ex(barcode_combo[rowid,1],barcode_combo[rowid,2])
  if(val<mismatch+1){
    error_message = paste0(error_message, barcode_combo[rowid,1]," to ", barcode_combo[rowid,2], " has a difference of ",val, " \n")
  }
}

#if there were errors, print message and exit
if(error_message!=""){
  cat(paste0("The number of differences between barocdes is less than, or equal to, the number of mismatches in the following barcodes:\n",error_message))
} else{
  print("Barcode schema passes QC")
}

#Create mutant lists for each 
nuc_list=c("A","C","T","G","N")
for (bc in bc_exp){
  mut_list=""
  
  #for each position of string
  for(bc_pos in 1:length(strsplit(bc,split="")[[1]])){
    
    #vary by nucleotide list
    for(nuc in nuc_list){
      bc_mut = bc
      substr(bc_mut, bc_pos, bc_pos) <- nuc
      
      #save variation to mut_list
      mut_list = c(mut_list,bc_mut)
    }
  }
  #add expected bc
  mut_list=c(mut_list,bc)
  
  #assign list to bc variable name
  assign(bc,mut_list[!duplicated(mut_list)])
}

#Read through input and assign barcodeID
df_counts$barcode_id=df_counts$barcode
for (rowid in 1:nrow(df_counts)){
  #assign barcode observed
  bc_obs = df_counts[rowid,"barcode"]
  
  #for each of the barcodes_expected
  for(bc in bc_exp){

    #if the expected barcode is in that list, assign ID as expected
    #otherwise leave as new barcode
    if(bc_obs %in% get(bc)){
      df_counts[rowid,"barcode_id"]=paste0("*",bc)
    } else{
      next
    }
  }
}

#merge df
df_counts_merged = aggregate(counts~barcode_id, df_counts, sum)

#sort and subset
df_counts_merged = df_counts_merged[order(df_counts_merged$counts,decreasing = TRUE),][1:number_of_barcodes_observed,]

#compare observed bc list with expected bc list, write output to text
diff_lis=setdiff(paste0("*",bc_exp),df_counts_merged$barcode_id)

# if the expected list matches with top number_of_barcodes_observed, print text, otherwise print error messages
# if the expected list matches with top number_of_barcodes_observed, print plot
if(length(diff_lis)==0){
  print("Barcodes observed pass QC")
  #print text file
  file_save = paste0(output_dir,'/',mpid,'_barcode_passed.txt')
  fileConn<-file(file_save)
  line1 = paste0("\n* SampleID ",mpid, " \n")
  line2 = paste0("\t + Number of mismatches allowed ", mismatch, " \n")
  line3 = paste0("\t + The top barcodes identified (", paste(df_counts_merged$barcode_id,collapse = ' '),
                 ") include the expected barcodes (", paste(bc_exp,collapse = ' '), ")")
  line4 = "\t + List of top barcodes:counts \n"
  line5 = ""
  for(rowid in 1:nrow(df_counts_merged)){
    line5=paste0(line5,"\t\t + ", df_counts_merged[rowid,"barcode_id"],": ", df_counts_merged[rowid,"counts"], "\n")
  }
  
  writeLines(paste(line1,line2,line3,line4,line5,sep=""), fileConn)
  close(fileConn)
  
  #print plot
  file_save = paste0(output_dir,'/',mpid,'_barcode.png')
  p = ggplot(df_counts_merged, aes(x = reorder(barcode_id,-counts),counts)) +
    geom_bar(stat ="identity") +
    geom_text(aes(label = counts), vjust = -0.2, size=2)+
    labs(title=paste0("Barcode Counts for ", mpid), x ="Barcode IDs", y = "Barcode Counts") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=5))
  ggsave(file_save,p)
  
} else{
  file_save = paste0(qc_dir,'/',mpid,'_barcode_errors.txt')
  fileConn<-file(file_save)
  line1 = paste0("\n* SampleID ",mpid, " \n")
  line2 = paste0("\t + Number of mismatches allowed ", mismatch, " \n")
  line3 = paste0("\t + The top barcodes identified (", paste(df_counts_merged$barcode_id,collapse = ' '),
                 ") were not congruent with the expected barcodes (", paste(bc_exp,collapse = ' '), ") \n")
  line4 = "\t + List of top barcodes:counts \n"
  
  line5 = ""
  for(rowid in 1:nrow(df_counts_merged)){
    line5=paste0(line5,"\t\t + ", df_counts_merged[rowid,"barcode_id"],": ", df_counts_merged[rowid,"counts"], "\n")
  }
  
  writeLines(paste(line1,line2,line3,line4,line5,sep=""), fileConn)
  close(fileConn)

  print(paste0("Barcodes failed QC check. Review file located here:",file_save,"."))
  print(paste0("To check the reads go to the /project/01_preprocess/01_fastq/ dir and run `zcat {name_of_sample} | wc -l. Divide this number by 4 to get the total number of reads"))
  print(paste0("If satisfied with this read count change the config/snakemake_config file barcode_qc_flag to IGNORE"))
}
